### PR TITLE
Add water models

### LINKED ISSUE #4: Water and ions
At one point, we had separate FF files floating around for water (TIP3P) and ions. Currently, this does not seem to be the case (at least not in this repo) which means that someone applying our FF to a full system (e.g. a host-guest system for example) that is solvated will likely end up with flexible force field water and, I THINK, no parameters for their ions. 

We should probably provide auxiliary libraries here for ions and TIP3P. (When we switch to another water model that should also be provided.)

### PR DESCRIPTION
**Pull request by mattwthompson**
Partially resolves #4

- [x] Make subdirectory of water models discoverable
- [x] Add water models (with or without Joung & Cheatham ion parameters)
  - [x] TIP3P + JC
  - [x] TIP3P-FB
- [x] Tests
  - [x] Versioning guarantee
  - [x] Comparison (with non-zero tolerances) to OpenMM parameters
    - [x] TIP3P
    - [x] TIP3P-FB
- [x] Add force field generation scripts
  - [x] TIP3P + JC
  - [x] TIP3P-FB
- [x] User-facing documentation
  - [x] Explain versioning scheme
  - [x] Guarantee that short-hand file matches most recent versioned file
  - [x] Disclaim differences associated with using short-hand file
  - [x] Document source of each parameters
  - [x] Document generation of each file

### CODE REVIEW

**Comment by mattwthompson** [comment]
Using TIP3P parameters I coped from an ambiguous source, I get the following parameter differences compared to OpenMM's `tip3p.xml`:

```
E           openff.toolkit.tests.utils.FailedParameterComparisonError: A difference between particle was detected. Details follow.
E
E           The following particles have different parameters in the two NonbondedForce forces:
E           particle 1:
E               epsilon: 0.635968 kJ/mol != 0.6363864000000001 kJ/mol
E               sigma: 0.3150752406575124 nm != 0.31506999999999996 nm
```

The 1983 TIP3P paper uses the following formulation of the 12-6 Lennard-Jones potential

image

with these parameters

image

So $A = 4 \epsilon \sigma^{12}$, $C = 4 \epsilon \sigma^{6}$, $\sigma = (A / C)^{1/6}$, $\epsilon = A / (4 \sigma^{12}) = C/ (4 \sigma^{6})$

```python3
>>> from openff.units import unit, Quantity
>>> A = Quantity(582.0 * 1e3, unit.kilocalorie_per_mole * unit.angstrom ** 12)
>>> C = Quantity(595.0, unit.kilocalorie_per_mole * unit.angstrom ** 6)
>>> sigma = (A / C) ** (1/6)
>>> epsilon = A / (4 * sigma ** 12)
>>> sigma
<Quantity(3.15065611, 'angstrom')>
>>> epsilon.to(unit.kilojoule_per_mole)
<Quantity(0.636271735, 'kilojoule_per_mole')>
```
These values, unfortunately, appear to be a third set of parameters if caring about the fifth or sixth digit of precision.

**Comment by mattwthompson** [comment]
A bit of confusion about the canonical TIP5P parameters: 

https://github.com/openmm/openmm/blob/c19b6d938a7a9e5f5909c27d92cf20db7a037e35/wrappers/python/openmm/app/data/tip5p.xml#L27

https://github.com/gromacs/gromacs/blob/9095f38ce5d9bb1627e2796f7e6fef1e9163cfe9/share/top/amber99.ff/ffnonbonded.itp#L66-L68

These are consistent but the files we've been throwing around use different parameters. I'm switching to using these but we need a stronger paper trail.

**Comment by j-wags** [comment]
So currently the big two things I'll want to see before I approve this are tests passing and something like the documentation-in-the-author-field thing from my last post.

**Comment by mattwthompson** [comment]
@lilyminium I wonder if you have any thoughts here, particularly [this](https://github.com/openforcefield/openff-forcefields/pull/57#issuecomment-1289566862) comment?

My preference is to decide on a single set of values, provided they are close to the other distributed versions of TIP3P, and visibly document how the files were generated and how some numerical values differ from others.

I grant that this is not the absolutely highest standard of precision in reproducibility. However, I think it bears emphasis that
* We're looking at numbers that differ around 4-6 significant digits. These probably matter but they clearly do not matter enough for everybody to use the same values.
* The discrepancies between original papers, OpenMM, and other sources do not seem to have caused a stir in the community
  * I spent some time looking through the issue tracker and couldn't find anybody raising a fuss over the numerical differences I found. This only an approximation of how much people care about these differences
* We were _already_ [exposing](https://github.com/openforcefield/openff-toolkit/issues/1538) a completely un-validated version of TIP3P and many other files due to a bug in the toolkit.
* If, for whatever reason, users want their own flavor of TIP3P, distributing these files provides a less error-prone starting point than having them write one from scratch or pull a less-validated file from the first Google result.


I'm not a fan of having `"tip3p_amber_1.0.0.offxml"`, `"tip3p_openmm_1.0.0.offxml"`, `"tip3p_amber_1.0.0.offxml"`, and `"tip3p_1983_jorgensen"` floating around in `openforcefield.smirnoff_forcefield_directory`; users just want to load a force field and get on with their work and (in my opinion) burdening them with these choices is no clear improvement over telling them to write their own file.

I think it's a matter of balancing rounding errors with the accessibility of having these force fields distributed in our infrastructure. I feel strongly that the benefit of being able to simply do `ForceField("tip3p_1.0.0.offxml")` outweighs the risk of 0.6363864 kJ/mol vs. 0.636271735 having an unambiguously correct value - and if it does, we can just patch the file and release a new version. In such a case, fixes would actually be propagated out to user's machines, an improvement over passing around the same slightly-incorrect file for years.

**Comment by lilyminium** [comment]
> I'm not a fan of having "tip3p_amber_1.0.0.offxml", "tip3p_openmm_1.0.0.offxml", "tip3p_amber_1.0.0.offxml", and "tip3p_1983_jorgensen" floating around in openforcefield.smirnoff_forcefield_directory; users just want to load a force field and get on with their work and (in my opinion) burdening them with these choices is no clear improvement over telling them to write their own file.

I definitely agree with this, tip3p parameters seem to vary all over the place -- GROMACS slightly different numbers in its amber force fields, for example. In the absence of a clear winner I would preference the original paper values, but I defer to you here.

> if it does, we can just patch the file and release a new version. In such a case, fixes would actually be propagated out to user's machines, an improvement over passing around the same slightly-incorrect file for years.

I don't 100% follow this logic though. Would the patched file still be called `tip3p_1.0.0.offxml` and users must specify `tip3p_1.0.1.offxml`, or are you proposing that using a `tip3p.offxml` is a shortcut for the latest, highest version file? If it's the former, I'm not sure I understand the practical difference between `tip3p_1.0.1.offxml` and `tip3p_amber_1.0.1.offxml`.

**Comment by mattwthompson** [comment]
Thanks for the feedback - you're right, that's a contradiction. I see three options being thrown around here

1. `tip3p.offxml`
2. `tip3p_x.y.z.offxml`
3. `tip3p_source_x.y.z.offxml`

that exist on a continuum between ease of use and explicitness in provenance. I consider the third option a non-starter for the same reasons you agreed with, but I didn't consider until your comment that we can't have the ease-of-use of the first option and also the fix-through-patches feature of the second. 

The mainline force fields have this problem (a "get the most recent OpenFF force field" function has been asked for but never got off the ground) but those updates are highly visible and, by comparison to fixing a rounding error in a water model, something users are likely to pay attention to and care enough to modify their scripts.

> are you proposing that using a tip3p.offxml is a shortcut for the latest, highest version file? 

I hadn't considered this - Jeff suggested early on that the version should be included in the file name and I didn't think about anything else. Thinking about it only briefly, I don't mind this idea since updates to these files should be rare[^1] and small[^2] and I think it matches users' soft expectations of provided water models[^3]. It's not strictly ideal that `ForceField("tip3p.offxml")` or similar might return (slightly!) different values in different years, but we can track this in release notes alongside versions[^4] of this package.

I'd like to give it a few more days of thought, but I like the idea that
* Models will be shipped as `"tip3p.offxml"`, `"tip4p-ew.offxml"`, etc.
* When changes clearly must be made, the same files will be modified and new releases (using CalVer) will be made
* There must be thorough documentation explaining changes, how values were calculated, etc. either in release notes or in another document linked to in release notes

I might change my mind tomorrow, but how does this sound to you?

[^1]: as rare as never, for most models, I hope
[^2]: if the work I've already done here is right and the references I followed are also right, differences should be numerically small
[^3]: analogous behavior with the mainline force fields would be bad since there are some tangible differences between major versions ... I think if for some reason somebody really cares about a specific flavor of TIP3P ~~they likely have bigger problems~~ they can of course use their own local file with preferred parameter values
[^4]: we committed to changing this package's versioning to CalVer, but the current coupling to SemVer wouldn't be able to capture bumping one of these models from i.e. 1.0.2 to 1.0.3

**Comment by lilyminium** [comment]
> Thinking about it only briefly, I don't mind this idea since updates to these files should be rare[1](https://github.com/openforcefield/openff-forcefields/pull/57#user-content-fn-1-b8311372a4c1828eed7dab7a20db51bd) and small[2](https://github.com/openforcefield/openff-forcefields/pull/57#user-content-fn-2-b8311372a4c1828eed7dab7a20db51bd) and I think it matches users' soft expectations of provided water models

Yes, I agree -- I like the simplicity of "tip3p.offxml" and also don't expect values to change much or significantly over the years. Would it be possible to double dip on options 1 and 2, and have both a "tip3p_1.0.0.offxml" and "tip3p.offxml" (that are actually the same file) in case users need to go back to older values over time, or prefer to specify the exact version?

> they can of course use their own local file with preferred parameter values

This might be more work than expected for people, compared to shipping versioned files with the repo.

**Comment by mattwthompson** [comment]
> Would it be possible to double dip on options 1 and 2, and have both a "tip3p_1.0.0.offxml" and "tip3p.offxml" (that are actually the same file) 

I think so ... we might want to add in some automation that enforces this policy (until we don't like it and turn it off).

> [users modifying local files] might be more work than expected for people, compared to shipping versioned files with the repo.

I think that's a feature of this idea - there's no way we can be responsible for "wrong" parameters if the users modify them to their favorite values. No need for a disclaimer check box in the case of a user wanting their own version of TIP3P that differs slightly from whatever version we ship.

**Comment by lilyminium** [comment]
> I think that's a feature of this idea - there's no way we can be responsible for "wrong" parameters if the users modify them to their favorite values. 

I agree with that. The more I think about it the more I like combining both `tip3p.offxml` and `tip3p_x.y.z.offxml` so users *can* pin to a water version if they want, without having to pin an `openff-forcefields` version that might not have other parameters they need -- or alternatively just use the most up-to-date version without thinking much about it.

**Comment by mattwthompson** [comment]
I'd like to shove all of these files in the same directory. Another option is to OpenMM's convention of requiring the `water/` prepend, but this does not work with the current toolkit. It [recursively globs](https://github.com/openforcefield/openff-toolkit/blob/7246056fa17332c793e53a8fe3276b0a07baf63c/openff/toolkit/typing/engines/smirnoff/forcefield.py#L123) all OFFXML files in a path. As a result, `'file.offxml'` and `'dir/file.offxml'` will both be exposed, even though `get_available_force_fields` returns only the file name without the path.

More concretely: before 5575e7fedaf3d0243e326f9b1f2819f81ecfb322, `len(get_available_force_fields())` and `len(set(get_available_force_fields()))` would not be equal and it's not clear how the collisions would be dealt with at force field loading time.

If we want these to be available _only_ as `ForceField("water/tip3p_1.0.0.offxml")` and have `ForceField("tip3p_1.0.0.offxml")` raise an exception, this will require a behavior change in the toolkit.

**Comment by mattwthompson** [comment]
Okay, I'm opening this up for review. This does not implement everything I set out to do originally, but I think it constitutes a large feature improvement and sets the stage for subsequent additions without wading into slightly murky waters.

* I dropped the virtual site models out of this PR (but [stashed](https://gist.github.com/mattwthompson/80852f252c02cced93ff09286378c31c) them). I'd like to tackle them in a follow-up, but I'd also like more time to go through the Interchange outputs with a fine-tooth comb. (The highly specific per-parameter testing is also a bit of work to put together to the level of safety we need and that partially motivates why I want to put on this.) Given who we understand the median user of OpenFF force fields to be, I think this is a worthwhile tradeoff.


I'm open to/soliciting feedback on

* I'm not sure exactly how the documentation should be structured here. The minimum information here is a lot and I'm not sure continuously appending to a monolithic README file is what we want to do long-term.
* I'm tracking the scripts I used to generate the OFFXML files. I think this is a small improvement over just shipping the files. I can either double down on this and add some automation for how these files should be updated or drop them altogether.

**Comment by lilyminium** [comment]
Thanks for writing this documentation, @mattwthompson. It doesn't look overly long to me -- do you anticipate that it will change or be added to much? Other than readable document I'm not sure what alternatives make sense here -- adding module docstrings to the scripts you use to generate the offxmls seems like the same thing, but harder. 

Tracking the scripts is a great idea! However, should they be shipped with the released package? IMO very few people will ever use them.

If you wanted to ensure the scripts remained in date, you could possibly auto-generate the final `offxml` files as part of CI.

**Comment by mattwthompson** [comment]
I don't anticipate any changes to the docs, or at least from my end. If changes are asked for in review I'll of course consider those.

> you could possibly auto-generate ...

That's what I was thinking in the "lean into automation" direction - I'll update CI to mirror what I did locally to make these files. I'll try to make sure the scripts aren't globbed up in the package data, too.

File: docs/water-models.md
Comment (j-wags) [inline docs/water-models.md:14]: (not blocking) I don't think this is necessary in this documentation, and is one more contemporary statement that can fall out of date. 

```suggestion
```

**Review by j-wags — COMMENTED** [review]
* I agree with the choice to not have variants for different `_source_`s and instead just package a single one
* I think this could be a good lower-stakes way to test a "latest" version convention with `tip3p.offxml`, and I really like that CI actively ensures it matches the highest version
* (not blocking) it would be really nice to somehow tell the user "you should probably be loading this FF last otherwise its parameters could get overwritten by eg. the water model in sage"
* I agree with the choice to just do tip3p initially
* I agree with the choice to keep the directory hierarchy flat (my reasoning is that FF jurisdictions already aren't clearly defined, eg this should really be `water_and_ions/tip3p.offxml` and things will only get more confusing from there)
* (blocking) I see in Sage that some free floating ions have their [SMIRKS decorated with `X0`](https://github.com/openforcefield/openff-forcefields/blob/0cc84eed2ca63b9adc9ab5ae4f5e5f1d02cfd3f5/openforcefields/offxml/openff-2.0.0.offxml#L369). This could be a good idea for all ions here, since we may eventually get a user report for vdW parameter misassignment to [hypervalent iodine](https://www.organic-chemistry.org/Highlights/2005/25May.shtm) or some other weird-but-not-utterly-unreasonable compound. 

Overall I'm happy with this PR and think it's basically "good enough", though we could continue improving it forever. 

@lilyminium - I'm technically the product owner for openff-forcefields, so to clarify any ambiguity for @mattwthompson: Only my approval is required to merge this. However Lily's thought WAY more about this recently than I have, so I'd like to know if you see any of your current comments/remaining thoughts as blocking.

File: scripts/write_tip3p.py
Comment (mattwthompson) [inline scripts/write_tip3p.py:77]: This logic could be simplified by storing SMIRKS patterns in the CSV file directly, but I value the human-readability of "Cs+" there.

I can't say for sure this logic is bulletproof, but the toolkit rejected some invalid patterns on earlier iterations of this code and the result matches (to my eye) what's in Sage.

File: scripts/write_tip3p.py
Comment (mattwthompson) [inline scripts/write_tip3p.py:69]: Please don't rubber-stamp these lines; I may have a PhD but I'm liable to misread and/or mistype short strings and ints.

**Comment by mattwthompson** [comment]
With any luck, cfc17a1 handles the logic of what I wanted to do: regenerate the OFFXML files and error if the files appear to differ. Let's see if it throws an error if I intentionally introduce changes ...

This might catch some false positives if minor changes happen in the toolkit, I think that's a bridge to cross when we get there. What I'm fairly confident it will do is complain if we introduce `tip3p_1.0.1.offxml` via the script and _don't_ track it. As it stands, making a new version of one of the models requires changing the `VERSION` variable and modifying the script with whatever intended changes. This could possibly be streamlined but it should be rare and I'm not eager to hyper-optimize that process.


It will fail when Rosemary or Sage 2.1 drops; I think this can be fixed with some regex but I don't have the juice to figure that out at the moment. (I think regex that does "running `git diff` on`openforcefields/offxml/` but exclude `openforcefields/offxml/^openff`" would handle it well enough.)

**Comment by mattwthompson** [comment]
Awesome, 97e69e2 confirms that changes to the files will fail this CI setup.

File: scripts/write_tip3p.py
Comment (j-wags) [reply scripts/write_tip3p.py:69]: Checked and verified!

**Review by j-wags — APPROVED** [review]
I see there's still a [DNM] in the title, but I don't have any more comments to add here, and it looks like this satisfies the in-scope asks from previous meetings, issues, and the discussion on this PR. This looks good to me, and can be merged when you're ready, @mattwthompson.

**Pull request merged**
